libraries/stdlib/osl/stdosl.h

```diff
@@ -654,6 +689,10 @@ closure color subsurface_bssrdf(normal N, color albedo, float sss_depth, color s
 // This closure may be vertically layered over a base BSDF, where energy that is not reflected
 // will be transmitted to the base closure.
 //
+//  \param  N           Normal vector of the surface point beeing shaded.
+//  \param  albedo      Surface albedo.
+//  \param  roughness   Surface roughness [0,1].
+//
 closure color sheen_bsdf(normal N, color albedo, float roughness) BUILTIN;
 ​
 // Adds an iridescent thin film layer over a microfacet base BSDF. This must be layered over
```

Comment:
I think requiring the use of the `layer` closure will potentially make the implementation rather awkward. Having extra arguments to the 3 closures that accept it would be easier (they could be optional arguments).

But maybe its better to wait until we can attempt the implementation in `testrender` to decide.

Reply:
I'm ok with relying on layer(). Our latest implementation is moving away from nested closures because GPU doesn't like recursion. The layer() model would be more convenient. The way we did it might not be mature enough, but I think layer() gives you enough freedom. You could even keep nesting recursively if you wanted.

On the other hand, nesting closures makes it more cumbersome if you want to do it without recursion. What did you have in mind?

Reply:
I think Chris was only referring to the implementation of thin film here. Since thin film is not a real closure, and more of a closure modifier (modifying Fresnel), implementing this as optional parameters for the closures that supports it would probably be easier than implementing it through the layer operator (as noted in the TODO comment below).

This is what we do in Arnold on our side, if thickness is set and > 0.0 the thin film Fresnel effect is enabled.

Reply:
Oh, right